NuRadioReco/modules/neutrinoEnergyReconstructor.py

```diff
@@ -0,0 +1,120 @@
+from NuRadioReco.utilities import trace_utilities, units
+
+
+class neutrinoEnergyReconstructor():
+    """
+    Module to perform an energy reconstruction as described in https://arxiv.org/abs/2107.02604
+    Look at NuRadioReco/examples/RNO_energy_reconstruction for an example how to use it.
+    Requires vertex reconstruction and IFT electric field reconstruction to be run beforehand.
+    """
+    def __init__(self):
+        self.__s_parametrization = np.array([4.70, -22.20, 29.99])
+        self.__s_prime_parametrization = np.array([6.46, -16.00, 12.49])
+        self.__s_passbands = np.array([
+            [.13, .3],
+            [.3, .5]
+        ])
+        self.__s_prime_passbands = np.array([
+            [.13, .2],
+            [.2, .3]
+        ])
+        self.__s_s_prime_cut_value = 10.
+        self.__channel_groups = None
+        self.__raytracer = None
+
+    def begin(self, channel_groups, ice, attenuation_model='GL1'):
+        """
+        Set up module
+
+        Parameters
+        -------------------
+        channel_groups: 2D array of integers
+            List of groups of channels which were used together by the iftElectricFieldReconstructor
+        ice: ice model object
+            Class describing the index of refraction model of the ice
+        attenuation model: string
+            Name of the model describing the signal attenuation
+        """
+        self.__channel_groups = channel_groups
+        self.__raytracer = NuRadioMC.SignalProp.analyticraytracing.ray_tracing(
+            ice,
+            attenuation_model
+        )
+        pass
+
+    def run(self, event, station, detector):
+        if not station.has_parameter(stnp.nu_vertex):
+            raise ValueError('Cannot run energy reconstruction without a reconstructed vertex!')
+        rec_vertex = station.get_parameter(stnp.nu_vertex)
+        for i_group, channel_group in enumerate(self.__channel_groups):
+            max_snr_in_group = 0
+            self.__raytracer.set_start_and_end_point(rec_vertex, detector.get_relative_position(station.get_id(), channel_group[0]))
+            self.__raytracer.find_solutions()
+            for i_efield, efield in enumerate(station.get_electric_fields()):
+                ray_type = efield.get_parameter(efp.ray_path_type)
+                channel = station.get_channel(channel_group[0])
+                # Check if this is the ray type with the largest SNR
+                if not channel.has_parameter(chp.signal_region_snrs) or not channel.has_parameter(chp.signal_ray_types):
+                    continue
+                current_snr = 0
+                for i_region, region_snr in enumerate(channel.get_parameter(chp.signal_region_snrs)):
+                    if ray_type == channel.get_parameter(chp.signal_ray_types)[i_region]:
+                        current_snr = region_snr
+                if current_snr <= max_snr_in_group:
+                    continue
+                max_snr_in_group = current_snr
```

Comment:
I'm not sure if this does what it should do - for one event, I got the following error:
```
Traceback (most recent call last):
  File "T04_energy_reconstruction.py", line 45, in <module>
    energy_reconstructor.run(event, station, det)
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/neutrinoEnergyReconstructor.py", line 121, in run
    event.add_shower(new_shower)
  File "/home/sb/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py", line 250, in add_shower
    raise AttributeError(f"shower with id {shower.get_id()} already exists. Shower id needs to be unique per event")
AttributeError: shower with id 0 already exists. Shower id needs to be unique per event
```
It may not be due to this code here, but I think if the first efield does not have the highest SNR, it will end up running twice and therefore attempt to add a shower with the same `shower_id` twice, raising this error?

Reply:
the easiest solution is probably to just unindent lines 118-120 below that create and add the shower, assuming the rest of the code below does what it's supposed to.